src/SelfUpdateCommand.php

```diff
@@ -101,38 +127,54 @@ protected function getReleasesFromGithub()
     }
 
     /**
-     * Get latest release according to given constraints
+     * Get the latest release version and download URL according to given constraints.
+     *
+     * @throws \Exception
+     *
+     * @return string[]|null
+     *  "version" and "download_url" elements if the latest release is available, otherwise - NULL.
      */
-    public function getLatestReleaseFromGithub($preview = false, $major_constraint = '') {
-        $releases = $this->getReleasesFromGithub();
-        $version = null;
-        $url = null;
-
-        foreach ($releases as $release) {
+    public function getLatestReleaseFromGithub() {
+        foreach ($this->getReleasesFromGithub() as $release) {
             // We do not care about this release if it does not contain assets.
-            if (count($release['assets']) && is_object($release['assets'][0])) {
-                $current_version = $release['tag_name'];
-                if ($major_constraint) {
-                    if (!Semver::satisfies($current_version, $major_constraint)) {
-                        // If it does not satisfies, look for the next one.
-                        continue;
-                    }
-                }
-                if (!$preview && VersionParser::parseStability($current_version) !== 'stable') {
-                    // If preview not requested and current version is not stable, look for the next one.
-                    continue;
-                }
-                $url = $release['assets'][0]->browser_download_url;
-                $version = $current_version;
-                break;
+            if (!isset($release['assets'][0]) || !is_object($release['assets'][0])) {
+                continue;
+            }
+
+            $releaseVersion = $release['tag_name'];
+            if ($this->isCompatible && !$this->satisfiesMajorVersionConstraint($releaseVersion)) {
+                // If it does not satisfies, look for the next one.
+                continue;
+            }
+
+            if (!$this->isPreview && VersionParser::parseStability($releaseVersion) !== 'stable') {
+                // If preview not requested and current version is not stable, look for the next one.
+                continue;
             }
+
+            if (null !== $this->versionConstraint && !Semver::satisfies($releaseVersion, $this->versionConstraint)) {
+                // Release version does not match version constraint option.
+                continue;
+            }
+
+            if (Semver::satisfies($releaseVersion, $this->currentVersion)) {
+                // The latest release matches the current one.
+                return null;
+            }
+
+            return [
+                'version' => $releaseVersion,
+                'download_url' => $release['assets'][0]->browser_download_url,
+            ];
         }
 
-        return [ $version, $url ];
+        return null;
```

Comment:
Stylistically I prefer having a uniform return type, and leave the current version check outside of this method. Optional.

Reply:
The current version check moved outside of the method